File: homecontroller_api/src/util/logger.ts
class LoggerImpl {
    private masterPrefix: string;
    private logPrefix: string;
    private errPrefix: string;
    private cshPrefix: string;
    private dbgPrefix: string;

    private debugEnabled: boolean;

    constructor() {
        this.masterPrefix = "[hc]";
        this.logPrefix = "[LOG]   ";
        this.errPrefix = "[ERR]   ";
        this.cshPrefix = "[FATAL] ";
        this.dbgPrefix = "[DEBUG] ";

        this.debugEnabled = false;
    }

    public enableDebug(): void {
        this.debugEnabled = true;
    }

    public disableDebug(): void {
        this.debugEnabled = false;
    }

    public log(msg: string, nl: boolean = false): void {
        this.print(this.logPrefix, msg, nl, '');
    }

    public err(msg: string, nl: boolean = false): void {
        this.print(this.errPrefix, msg, nl, '\x1b[0;93m');
    }

    public csh(msg: string, nl: boolean = false): void {
        this.print(this.cshPrefix, msg, nl, '\x1b[0;91m');
    }

    public dbg(msg: string, nl: boolean = false): void {
        if (this.debugEnabled) {
            this.print(this.dbgPrefix, msg, nl, '\x1b[3;90m');
        }
    }

    private timestamp(): string {
        const currDate = new Date();

        const dateStr: string = (currDate.getMonth() + 1) + '-' + currDate.getDate().toString().padStart(2, '0') + '-' + currDate.getFullYear().toString().substring(2);
        const timeStr: string = currDate.getHours().toString().padStart(2, '0') + ':' + currDate.getMinutes() + ':' + currDate.getSeconds().toString().padStart(2, '0');

        return dateStr + ' ' + timeStr;
    }

    private print(prefix: string, msg: string, nl: boolean, color: string): void {
        let outMsg: string = color + this.masterPrefix + ' ' + this.timestamp() + ' ' + prefix + '\t' + msg;
        
        if (nl) {
            outMsg += '\n';
        }

        outMsg += '\x1b[0m';

        console.log(outMsg);
    }
}

export const Logger = new LoggerImpl();

File: libhomecontroller/src/util/logger.cpp
#include "homecontroller/util/logger.h"

#include <ctime>
#include <sstream>
#include <iomanip>
#include <iostream>

namespace hc {
namespace util {

    const std::string logger::_MASTER_PREFIX = "[hc]";

    const std::string logger::_LOG_PREFIX = "[LOG]   ";
    const std::string logger::_ERR_PREFIX = "[ERR]   ";
    const std::string logger::_CSH_PREFIX = "[FATAL] ";
    const std::string logger::_DBG_PREFIX = "[DEBUG] ";

    bool logger::_DEBUG_ENABLED = false;

    void logger::log(const std::string& msg, bool nl) {
        print(_LOG_PREFIX, msg, nl);
    }

    void logger::err(const std::string& msg, bool nl) {
        print(_ERR_PREFIX, msg, nl, "\033[0;93m");
    }

    void logger::csh(const std::string& msg, bool nl) {
        print(_CSH_PREFIX, msg, nl, "\033[0;91m");
    }

    void logger::dbg(const std::string& msg, bool nl) {
        if (_DEBUG_ENABLED) print(_DBG_PREFIX, msg, nl, "\033[3;90m");
    }

    std::string logger::timestamp() {
        auto t = std::time(nullptr);

        std::ostringstream ss;
        ss << std::put_time(std::localtime(&t), "%m-%d-%y %H:%M:%S");

        return ss.str();
    }

    void logger::print(const std::string& prefix, const std::string& msg, bool nl, const std::string& color) {
        std::ostringstream ss;
        ss << color << _MASTER_PREFIX << " " << timestamp() << " " << prefix << "\t" << msg;
        if (nl) ss << "\n";
        ss << "\033[0m"; // default color

        std::cout << ss.str();
     }

}
}